File: android/app/src/main/java/com/roro/futurevoice/data/AppUsageLog.kt
package com.roro.futurevoice.data

import android.content.Context
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Foreground seconds per LOCAL day (`AppUsageLog.swift`) — the day card's
 * "study" figure. Written at the edges of a stint, and never less than the
 * talk figure when read: a call in a pocket is metered but not foregrounded.
 */
object AppUsageLog {
    private const val PREFS = "futurevoice"
    private const val KEY = "futurevoice.appSecondsByDay"
    private const val KEEP_DAYS = 45
    private var enteredAt = 0L

    fun begin() { enteredAt = System.currentTimeMillis() }

    fun end(context: Context) {
        val start = enteredAt
        if (start == 0L) return
        enteredAt = 0L
        val seconds = ((System.currentTimeMillis() - start) / 1000).toInt()
        if (seconds <= 0) return
        val map = load(context).toMutableMap()
        val k = dayKey(System.currentTimeMillis())
        map[k] = (map[k] ?: 0) + seconds
        save(context, prune(map))
    }

    fun secondsOn(context: Context, dayMillis: Long): Int =
        load(context)[dayKey(dayMillis)] ?: 0

    private fun dayKey(now: Long) = SimpleDateFormat("yyyy-MM-dd", Locale.US).format(Date(now))

    private fun prune(map: Map<String, Int>): Map<String, Int> {
        val cutoff = dayKey(System.currentTimeMillis() - KEEP_DAYS * 86_400_000L)
        return map.filterKeys { it >= cutoff }
    }

    private fun load(context: Context): Map<String, Int> =
        context.getSharedPreferences(PREFS, 0).getString(KEY, null)?.let { raw ->
            runCatching {
                raw.split(',').filter { it.contains('=') }.associate {
                    it.substringBeforeLast('=') to it.substringAfterLast('=').toInt()
                }
            }.getOrNull()
        } ?: emptyMap()

    private fun save(context: Context, map: Map<String, Int>) {
        context.getSharedPreferences(PREFS, 0).edit()
            .putString(KEY, map.entries.joinToString(",") { "${it.key}=${it.value}" }).apply()
    }
}

File: FutureVoice/Services/AppUsageLog.swift
import Foundation

/// Seconds the app spent in the FOREGROUND, per local day — "how long did I
/// study today", for the day card. Written at the edges of a foreground stint
/// (`FutureVoiceApp` on `scenePhase`), never polled.
///
/// Deliberately naive: a call that keeps running in a pocket is metered by
/// `TalkMeter` but not counted here, so the card takes `max(usage, talk)`. It
/// is a study-time figure, not a billing one — the day's talk seconds stay
/// with `TalkTimeLog`, which reads the meter.
@MainActor
enum AppUsageLog {
    private static let key = "futurevoice.foregroundSecondsByDay"
    private static let keepDays = 45
    private static var activeSince: Date?

    static func becameActive(now: Date = Date()) {
        activeSince = now
    }

    static func resigned(now: Date = Date()) {
        guard let since = activeSince else { return }
        activeSince = nil
        add(seconds: Int(now.timeIntervalSince(since)), now: now)
    }

    /// Stored seconds for a day — plus the open stint when the day is today,
    /// so a card made mid-session counts the minutes it is being made in.
    static func seconds(on day: Date, now: Date = Date()) -> Int {
        let stored = load()[dayKey(day)] ?? 0
        guard dayKey(day) == dayKey(now) else { return stored }
        let open = activeSince.map { max(0, Int(now.timeIntervalSince($0))) } ?? 0
        return stored + open
    }

    private static func add(seconds: Int, now: Date) {
        guard seconds > 0 else { return }
        var map = load()
        map[dayKey(now), default: 0] += seconds
        if map.count > keepDays {
            let cutoff = dayKey(now.addingTimeInterval(-Double(keepDays) * 86_400))
            map = map.filter { $0.key >= cutoff }
        }
        UserDefaults.standard.set(map, forKey: key)
    }

    private static func load() -> [String: Int] {
        UserDefaults.standard.dictionary(forKey: key) as? [String: Int] ?? [:]
    }

    /// LOCAL day, same shape as `TalkTimeLog`'s keys.
    static func dayKey(_ date: Date) -> String {
        let f = DateFormatter()
        f.calendar = .current
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyy-MM-dd"
        return f.string(from: date)
    }
}
